cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -537,6 +558,214 @@ aggregate_reader_metadata::aggregate_reader_metadata(
+
+  // Local arrow::schema object
+  arrow_schema_t schema;
+
+  // Read arrow schema from flatbuffers
+  std::string encoded_serialized_message = it->second;
+
+  // Decode the base64 encoded ipc message string
+
+  // Note: Store the output from base64_decode in the lvalue here and then pass
+  // it to decode_ipc_message. Directly passing rvalue from base64_decode to
+  // decode_ipc_message can lead to unintended nullptr dereferences.
+  auto decoded_message = cudf::detail::base64_decode(encoded_serialized_message);
+
+  // Decode the ipc message to get a const void pointer to the arrow:Message flatbuffer
+  auto metadata_buf = decode_ipc_message(decoded_message);
+
+  if (metadata_buf == nullptr) {
+    CUDF_LOG_ERROR("Parquet reader encountered an invalid metadata pointer.",
+                   "arrow:schema not processed.");
+    return std::nullopt;
+  }
+
+  // Get an arrow:ipc message object from ipc:message's decoded metadata section
+  auto ipc_message = flatbuf::GetMessage(metadata_buf);
+  if (ipc_message == nullptr) {
+    CUDF_LOG_ERROR("Parquet reader encountered an invalid ipc_message pointer.",
+                   "arrow:schema not processed.");
+    return std::nullopt;
+  }
+
+  // Get schema object from the arrow:ipc message
+  auto fb_schema = ipc_message->header_as_Schema();
+  if (fb_schema == nullptr) {
+    CUDF_LOG_ERROR("Parquet reader encountered an invalid fb_schema pointer.",
+                   "arrow:schema not processed.");
+    return std::nullopt;
+  }
+
+  // Get the vector of fields from arrow:schema object
+  const auto fields = fb_schema->fields();
+  if (fields == nullptr) {
+    CUDF_LOG_ERROR("Parquet reader encountered an invalid fields pointer.",
+                   "arrow:schema not processed.");
+    return std::nullopt;
+  }
+
+  // This lambda function converts from flatbuf::duration type to cudf::type_id
+  auto duration_from_flatbuffer = [](flatbuf::Duration const* duration) {
+    // TODO: we only need this for arrow::DurationType for now. Else, we can take in a
+    // void ptr and typecast it to the corresponding type based on the type_id.
+    auto fb_unit = duration->unit();
+    switch (fb_unit) {
+      case flatbuf::TimeUnit::TimeUnit_SECOND:
+        return cudf::data_type{cudf::type_id::DURATION_SECONDS};
+      case flatbuf::TimeUnit::TimeUnit_MILLISECOND:
+        return cudf::data_type{cudf::type_id::DURATION_MILLISECONDS};
+      case flatbuf::TimeUnit::TimeUnit_MICROSECOND:
+        return cudf::data_type{cudf::type_id::DURATION_MICROSECONDS};
+      case flatbuf::TimeUnit::TimeUnit_NANOSECOND:
+        return cudf::data_type{cudf::type_id::DURATION_NANOSECONDS};
+      default: break;
+    }
+    // 0 is simply a dummy value for the scalar
+    return cudf::data_type{};
+  };
+
+  // Question: Should we not walk the schema here and return flatbuf:Fields *
+  // and walk fields in `select_columns` function where we actually
+  // use the to_type_id function?
```

Comment:
That might be better, and would answer your question below about how to name the fields. You will definitely want to use a full path rather than just the field name.  For instance, if there's acolumn that's a list of durations, the leaf nodes will be named simply "element". If there are other list columns, that "element" name will collide.

A different approach might be to add an `arrow_type` field to the `SchemaElement` struct, so then the parquet and arrow schemas will align and we can use a single naming for nested columns. The mapping between arrow and parquet schemas looks to be pretty straightforward, but I'm just not sure about lists...in parquest list naming will look like
```
field1
  list
    element
```
with a `SchemaElement` for all three names (so the leaf column name is "field1.list.element"). I'm not sure about what you get out of decoding the flatbuffer, but looking at the schemas that pyarrow prints, the "list" node might not be there, but would have to be implied from the field type.  a pyarrow list looks like:
```
field1: list<element: string>
  child 0, element: string
```
So I'm not sure if the names will match...it seems like here the name would be "field1.element" :shrug:

Reply:
I am partial to just co-walking strategy when walking the parquet schema. Let me try that and push some changes to the PR. Thanks for the feedback Ed!

Reply:
As feared, the arrow and parquet schemas are nested differently.  The presence of lists throws a wrench in the `validate_schemas` lambda. Modifying the python test code to add a list segfaults. By the time you reach the bottom of the parquet schema tree, the arrow schema becomes null. I think you're going to have to throw a dummy node into the arrow tree when you encounter an arrow schema element of type 'list'.

```python
def test_parquet_reader_duration_types(tmpdir):
    pdf = pd.DataFrame(
        {
            "s": pd.Series([1234, 3456, 32442], dtype="timedelta64[s]"),
            "l": [[1,2],[1,2],[1,2]],
        }
    )
    # Write parquet with arrow for now (to write arrow:schema)
    fname = tmpdir / "duration.parquet"
    pdf.to_parquet(fname, engine="pyarrow")

    # Check results
    got = cudf.read_parquet(fname)
    assert_eq(pdf.dtypes, got.dtypes)
    assert_eq(pdf, got)
```

The following hack (around line 612 or so) worked around the segfault, but the test still fails because the list column types differ.  But at least "s" is a timedelta still :D

```c++
        if (flatbuffers::GetString(field->name()) == "item") {
          schema_elem.children.emplace_back(arrow_schema_data_types{});
          schema_elem.children.back().children = std::move(schema_children);
        } else {
          schema_elem.children = std::move(schema_children);
        }
```
```
E   AssertionError: Series are different
E   
E   Series values are different (50.0 %)
E   [index]: [s, l]
E   [left]:  [timedelta64[s], object]
E   [right]: [timedelta64[s], list]
E   At positional index 1, first diff: object != list
```

Reply:
FWIW, if `got` is converted to pandas (`got - cudf.read_parquet(fname).to_pandas()`) then the test passes.

Reply:
Thank you for investigating this and suggesting the dummy node fix, Ed! 

I am also looking into co-walk and I am going to make a slight modification to the logic you used. Instead of checking `name == item` which may or may not be true according to Arrow, I am going with `type == any of the list types` consistent with the writer side code of Arrow. This makes all my local tests (with structs, maps, lists) type columns pass at least. 

As for the `object != list` issue, that is simply arrow/cudf to pandas dtype interoperation thing and isn't really related to parquet or arrow schema so I am going to convert all tables to `cudf` before comparing them for an apple to apple comparison.

Example:

```python
def test_parquet_reader_arrow_schema_types():
    pdf = pd.DataFrame(
        {
            "s": pd.Series([1234, 3456, 32442], dtype="timedelta64[s]"),
            "ms": pd.Series([1234, 3456, 32442], dtype="timedelta64[ms]"),
            "us": pd.Series([1234, 3456, 32442], dtype="timedelta64[us]"),
            "ns": pd.Series([1234, 3456, 32442], dtype="timedelta64[ns]"),
        }
    )

    # Convert all pdf dtypes to int64
    pdf2 = pdf.astype("int64")

    # Write parquet with arrow for now (to write arrow:schema)
    buffer = BytesIO()
    pdf.to_parquet(buffer, engine="pyarrow")

    # Read parquet with arrow schema
    got = cudf.read_parquet(buffer)
    # Convert to cudf table for an apple to apple comparison
    expected = cudf.from_pandas(pdf)

    # Check results for reader with schema
    assert_eq(expected.dtypes, got.dtypes)
    assert_eq(expected, got)
```

Reply:
> Instead of checking `name == item`

Thank you 😄. That was for illustration only and horribly brittle...I was just too lazy to figure out how to cast the type appropriately 😅. I'm not sure what other nesting nightmares are lurking out there. Should probably work in some structs and maps as well to make sure we're not missing anything.

Reply:
Yep, I did structs and they are working fine. Surprisingly cudf doesn't support `maps` (aka categories) types with parquet so they don't round trip faithfully with or without schema but the code doesn't crash!

Reply:
Looks like the PR is in a good shape now to be reviewed!

Reply:
> Surprisingly cudf doesn't support `maps`

Weird...I wonder what's going on. I'm pretty sure I've seen references to maps (which are implemented as lists of key/value pairs IIRC). Follow on PR fodder I guess.

Reply:
Looking at it again, I think I am confusing between `map` and `category` types. Please correct me if I am wrong.

1. `category` type <=> cudf `DICTIONARY32` (categorial) <=> Arrow `DICTIONARY` <=> not code in cudf Parquet. Doesn't round trip faithfully.
2. `map` -> list of keys and values. Round trips ok.

```python
def test_parquet_arrow_schema():
    s = cudf.Series(['cat', 'dog', 'cow'])
    s.map({'cat': 'kitten', 'dog': 'puppy', 'cow':'calf'})
    expected2 = cudf.DataFrame(
        {
            "s": cudf.Series([1234, 123, 4123], dtype="timedelta64[s]"),
            "i64": cudf.Series([1234, 123, 4123], dtype="int64"),
            "l": list([[1,2],[1,2],[1,2]]),
            "time": cudf.Series([1234, 123, 4123], dtype="datetime64[ms]"),
            "cat" : cudf.Series(['a', 'b', 'a'], dtype=cudf.CategoricalDtype(categories=['b', 'a'], ordered=True)),
            "map": s,
        })

    # Read parquet with arrow schema
    #got = cudf.read_parquet(buffer)

    # Write parquet with arrow for now (to write arrow:schema)
    buffer2 = BytesIO()
    pdf2 = expected2.to_pandas()

    pdf2.to_parquet(buffer2, engine="pyarrow")
    pa_got = pd.read_parquet(buffer2, engine="pyarrow")
    # Read parquet with arrow schema
    got = cudf.read_parquet(buffer2)
    got2 = cudf.read_parquet(buffer2, use_arrow_schema=False)

    print("Original cudf table\n", expected2.dtypes)
    print("\nto_pandas\n", pdf2.dtypes)
    print("\ncudf.Parquet\n", got.dtypes)
    print("\ncudf.Parquet_no_schema\n", got2.dtypes)
    print("\npandas parquet\n", pa_got.dtypes)
```
prints:

```bash 
Original cudf table
 s       timedelta64[s]
i64              int64
l                 list
time    datetime64[ms]
cat           category
map             object
dtype: object

to_pandas
 s       timedelta64[s]
i64              int64
l               object
time    datetime64[ms]
cat           category
map             object
dtype: object

cudf.Parquet
 s       timedelta64[s]
i64              int64
l                 list
time    datetime64[ms]
cat             object
map             object
dtype: object

cudf.Parquet_no_schema
 s                int64
i64              int64
l                 list
time    datetime64[ms]
cat             object
map             object
dtype: object

pandas parquet
 s       timedelta64[s]
i64              int64
l               object
time    datetime64[ms]
cat           category
map             object
dtype: object
```

Reply:
@vuule  for vis